docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,303 @@
+pods and service accounts within a project
+
+
+## Requirements
+
+1.  Provide a set of restrictions that controlls how a security context is created as a new, cluster-scoped, object
+called SecurityContextConstraints.
+1.  User information in `user.Info` must be available to admission controllers. (Completed in
+https://github.com/GoogleCloudPlatform/kubernetes/pull/8203)
+1.  Some authorizers may restrict a user’s ability to reference a service account.  Systems requiring
+the ability to secure service accounts on a user level must be able to add a policy that enables
+referencing specific service accounts themselves.
+1.  Admission control must validate the creation of Pods against the allowed set of constraints.
+
+## Design
+
+### Model
+
+A security context constraint object exists in the root scope, outside of a namespace.  The
+security context constraints will reference users and groups that are allowed
+to operate under the constraints.  In order to support this, `ServiceAccounts` must be mapped
+to a user name or group list by the authentication/authorization layers.  This allows the security
+context to treat users, groups, and service accounts uniformly.
+
+Below is a list of security context constraints which will likely serve most use cases:
+
+1.  A default constraints object.  This object is permissioned to something covers all actors such
+as a `system:authenticated` group and will likely be the most restrictive set of constraints.
+1.  A default constraints object for service accounts.  This object can be identified as serving
+a group identified by `system:service-accounts` which can be imposed by the service account authenticator / token generator.
+1.  Cluster admin constraints identified by `system:cluster-admins` group - a set of constraints with elevated privileges that can be used
+by an administrative user or group.
+1.  Infrastructure components constraints which can be identified either by a specific service
+account or by a group containing all service accounts.
+
+```go
+// SecurityContextConstraints governs the ability to make requests that affect the SecurityContext that will
+// be applied to a container.
+type SecurityContextConstraints struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// AllowPrivilegedContainer determines if a container can request to be run as privileged.
+	AllowPrivilegedContainer bool `json:"allowPrivilegedContainer,omitempty" description:"allow containers to run as privileged"`
+	// AllowedCapabilities is a list of capabilities that can be requested to add to the container.
+	AllowedCapabilities []CapabilityType `json:"allowedCapabilities,omitempty" description:"capabilities that are allowed to be added"`
+	// AllowHostDirVolumePlugin determines if the policy allow containers to use the HostDir volume plugin
+	AllowHostDirVolumePlugin bool `json:"allowHostDirVolumePlugin,omitempty" description:"allow the use of the host dir volume plugin"`
```

Comment:
Which host dirs? What if I don't want to allow /var/run/docker.sock but I do want to allow /etc/ssl? What if I don't want to allow rw but I do want to allow r?